source/s3_client.c

```diff
@@ -930,253 +909,179 @@ static void s_s3_client_process_work_task(struct aws_task *task, void *arg, enum
     aws_linked_list_swap_contents(&vip_connections_updates, &client->synced_data.pending_vip_connection_updates);
-    aws_linked_list_swap_contents(&vip_connections_removals, &client->synced_data.pending_vip_connection_removals);
-    aws_linked_list_swap_contents(&meta_requests_adds, &client->synced_data.pending_meta_requests);
-
-    s_s3_client_unlock_synced_data(client);
 
-    /*******************/
-    /* Step 1: Process VIP connection removals, flagging each VIP connection for destruction */
-    /*******************/
-    while (!aws_linked_list_empty(&vip_connections_removals)) {
-        struct aws_linked_list_node *work_node = aws_linked_list_pop_back(&vip_connections_removals);
-        struct aws_s3_client_work *work = AWS_CONTAINER_OF(work_node, struct aws_s3_client_work, node);
-        struct aws_s3_vip_connection *vip_connection = work->user_data;
-
-        /* We can't necessarily clean up a VIP connection immediately--it may still be in the middle of processing a
-         * request. Instead, we flag it for pending destruction, and the next time it becomes idle, it will be
-         * destroyed. */
-        vip_connection->threaded_data.pending_destruction = true;
-
-        aws_linked_list_push_back(&work_destroy_list, &work->node);
-    }
+    client->threaded_data.num_requests_in_flight -= client->synced_data.pending_request_count;
+    client->synced_data.pending_request_count = 0;
 
-    /*******************/
-    /* Step 2: Process meta request adds. */
-    /*******************/
-    while (!aws_linked_list_empty(&meta_requests_adds)) {
-        struct aws_linked_list_node *work_node = aws_linked_list_pop_back(&meta_requests_adds);
-        struct aws_s3_client_work *work = AWS_CONTAINER_OF(work_node, struct aws_s3_client_work, node);
-        struct aws_s3_meta_request *meta_request = work->user_data;
-
-        /* If this is already scheduled, go ahead and clean it up. */
-        if (meta_request->threaded_data.added_to_client) {
-            aws_s3_meta_request_release(meta_request);
-        } else {
-            /* Push it to our list.  We don't call aws_s3_meta_request_acquire here since we still have the reference
-             * for when the meta request was added to the work list. */
-            aws_linked_list_push_back(&client->threaded_data.meta_requests, &meta_request->threaded_data.node);
-            meta_request->threaded_data.added_to_client = true;
-        }
-
-        aws_linked_list_push_back(&work_destroy_list, &work->node);
-    }
+    s_s3_client_unlock_synced_data(client);
 
     /*******************/
-    /* Step 3: Check all of the incoming VIP connections for meta requests.  If they already have a meta request, that
-     * means they just finished a request, so we check for remaining work.  If there isn't any remaining work, then we
-     * remove that meta request from our list. */
+    /* Step 2: Process any meta request work. */
```

Comment:
This section is much easier to follow that the old one, excellent!

Reply:
:-D